doc/stm/index.mld

```diff
@@ -0,0 +1,581 @@
+
+The [postcond] function is the last concerning the model against which the
+library will be tested. It takes a [cmd], the [state] {i before} running the
+[cmd] and the [res] returned by the [run] function. In this function, we
+express the relation between the result of the library's functions and the
+functional model.  The fact that it takes the [state] before the call of the
+function will come in handy when we will add a {{!section-remove}[remove]}
+function to our set library.
+
+{3 The [cmd]s generation}
+
+The last thing to do is to tell [qcheck-stm] how to generate the [cmd]s. This
+is what is done by the [arb_cmd] function, using [QCheck]'s combinators. We
+will make two remarks on this function.  The first one is that [QCheck.make]
+takes an optional printer. It is important to provide it so that the test's
+output is readable.  Here we use the function [show_cmd] that has been built by
+ppx_deriver from [cmd]'s definition.  The second remark is that [arb_cmd] takes
+a [state] as argument. This allows to make the generation of [cmd]s depend on
+the [state] of the model. We don't use it here, but again, it will be
+demonstrated when we will add a {{!section-remove}[remove]} function to our set
+library.
+
+
+{2 Running our first tests}
+
+Now, we are set to run our first [qcheck-stm] tests!
+
+We will begin with some sequential testing. {!module-STM_sequential} provides a
+functor that exposes mainly two functions:
+
+- {{!STM_sequential.Make.agree_test}agree_test} to build a positive [QCheck] test,
+- {{!STM_sequential.Make.neg_agree_test}neg_agree_test} to build a negative [QCheck] test.
+
+Here, we expect the test to succeed, so we will use the first one.
+
+{[
+module Lib_sequential = STM_sequential.Make (Lib_spec)
+
+let _ =
+  QCheck_runner.run_tests ~verbose:true
+    [ Lib_sequential.agree_test ~count:100 ~name:"STM sequential tests" ]
+]}
+
+And the test fails...
+
+{[
+$ dune exec ./mutable_set.exe
+random seed: 445566802
+generated error fail pass / total     time test name
+[✗]    2    0    1    1 /  100     0.0s STM sequential tests
+
+--- Failure --------------------------------------------------------------------
+
+Test STM sequential tests failed (11 shrink steps):
+
+[(Add -1397066193181869250); Cardinal]
+
++++ Messages ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+Messages for test STM sequential tests:
+
+  Results incompatible with model
+
+   (Add -1397066193181869250) : ()
+   Cardinal : 0
+================================================================================
+failure (1 tests failed, 0 tests errored, ran 1 tests)
+]}
+
+In the case of a failing test, [qcheck-stm] gives the user the counter example
+it has found, after some shrinking steps, as it is customary in property-based
+testing {i à la} QuickCheck. The counter example is a program. It is given
+twice: the first time just as a list of calls; the second time, each call is
+paired with the result of the computation.
```

Comment:
Should we fix `STM` so that the printing is more consistent as `cmd`s under each other (no `;` or square braces)?

Reply:
Yes, I think it would be nicer.

Reply:
I see you've created #218 for this and have a PR coming :+1: